File: 08/ruby/part02.rb
# Day 8: Space Image Format
# https://adventofcode.com/2019/day/8#part2

input = File.read('./inputs/08.txt').chomp.chars

WIDTH  = 25
HEIGHT =  6

flattened = Array.new(WIDTH * HEIGHT) { 2 }
while !input.empty?
  layer = input.take(WIDTH * HEIGHT)
  input = input.drop(WIDTH * HEIGHT)
  layer.each_with_index do |pixel, i|
    if flattened[i] == 2
      flattened[i] = pixel.to_i
    end
  end
end

while !flattened.empty?
  row = flattened.take(WIDTH)
  flattened = flattened.drop(WIDTH)
  pixels = row.map do |pixel|
    if pixel == 1
      '█'
    else
      ' '
    end
  end
  puts pixels.join('')
end

File: 08/nim/part02.nim
# Day 8: Space Image Format
# https://adventofcode.com/2019/day/8#part2

import algorithm, strutils, sequtils

let
  input = open("./inputs/08.txt")
  width = 25
  height = 6

var
  flattened = newSeq[char](width * height)
  pixels = toSeq(strip(input.readAll()))

input.close()
fill(flattened, '2')

proc switchChar(c: char): string =
  result = " "
  if c == '1':
    result = "█"
  return

var offset = 0
while offset < len(pixels):
  let layer = pixels[offset..(offset + (width * height) - 1)]

  for i, c in layer:
    if flattened[i] == '2':
      flattened[i] = c

  offset += width * height

for h in 0..(height - 1):
  let row = flattened[(h * width)..((h * width) + width - 1)]
  echo foldl(row, a & switchChar(b), "")
